### ISSUE TITLE
[Bug][ROCm] GLM-5.3-Flash kpool 32-page split (hybrid block_size=1152) GPU Memory access fault on MTP decode

### ISSUE THREAD

**Issue by zzw09773** [vllm-project/vllm#55280]
ROCm gfx942, GLM-5.3-Flash, hybrid KDA + MLA + kpool indexer: after the hybrid aligner lands on `cache_config.block_size = 1152`, the kpool indexer virtually splits 288-entry storage into **32-entry** paged-MQA pages. MTP decode then dies with `Memory access fault by GPU node-N` (one `VllmWorker` abort → engine down). Not OOM.

Line numbers below are **vLLM `main` @ `156050598eea0bb1124b9c106b22ba4a8c56f82a`** (2026-09-04). Runtime numbers are from a gfx942 serving image derived from #53906 plus later local patches; they are labelled as such.

## Environment (runtime)

- 8× MI325X (gfx942), ROCm 7.2.3
- GLM-5.3-Flash, TP=4, MTP `num_speculative_tokens=2`
- Hybrid KDA + MLA + kpool indexer (`index_kpool = 4`)
- Serving image: #53906-era tree. Crash B already included the V2 slot-mapping `in_range` guard from #54296.

## Symptom

Two serving aborts on 2026-09-03, same family:

- Crash A (03:41Z): ~41054 computed tokens, first MTP decode after that prefill. Image did **not** yet have the #54296 V2 guard.
- Crash B (09:18Z): 5-way mixed batch (4 MTP decode + one 2304-token chunked prefill). `new_block_ids` around **3809–3813**. Prefix-cache hit ~93%, KV usage ~4.25%. Image **already had** the #54296 V2 guard.

Banner is `Memory access fault by GPU node-N` → that rank’s `VllmWorker` exits → engine down. Not OOM. An aiter GEMM-shape miss in the same logs is not sufficient (same shape had been served earlier without abort). `scheduled_spec_decode_tokens=[-1,-1]` is the async placeholder, not the address.

## Why 1152 becomes a 32-page split

`PAGED_MQA_PAGE_SIZES = (32, 64)` (`vllm/utils/deep_gemm.py:35`).

`Glm5NextIndexerCache.get_kv_cache_spec` (`vllm/models/glm5next/nvidia/attention.py:145-152`; there is no `amd/attention.py` sibling) picks the virtual page as:

- native if `storage_block_size <= 64`
- 64 if `storage_block_size % 64 == 0`
- **else 32**

On this host the hybrid mamba aligner set `block_size = 1152`. With `index_kpool = 4`:

`1152 / 4 = 288`, `288 > 64` and `288 % 64 != 0` → **page_size = 32**.

That is allowed by the assert at `attention.py:136-137` (only requires a multiple of 32). It is exactly the fallback the comment at `attention.py:132-134` describes as “virtually split”.

## CUDA already avoids this on sm120; ROCm does not align at all

Hybrid padding is applied in `Platform._align_hybrid_block_size` (`vllm/platforms/interface.py:927-929`):

```python
indexer_align = cls._get_indexer_block_alignment(vllm_config)
if indexer_align:
    attn_block_size = indexer_align * cdiv(attn_block_size, indexer_align)
```

`CudaPlatform._get_indexer_block_alignment` (`vllm/platforms/cuda.py:414-431`) returns `index_kpool * min(PAGED_MQA_PAGE_SIZES)` (= 128 for kpool=4), except **sm120**, where it already uses `max` (= 256) so the split lands on 64 not 32. The comment at `cuda.py:427-429` is the same 32-vs-64 issue.

`RocmPlatform` (`vllm/platforms/rocm.py:498`) does **not** override `_get_indexer_block_alignment`. It inherits `Platform`’s default (`interface.py:767-775`), which returns `None` and documents that “Other platforms impose no extra constraint.” So on ROCm the `if indexer_align:` branch never runs, and hybrid GLM stays on 1152 → 32-page.

Even CUDA’s non-sm120 `min` path is a no-op for 1152 (`1152 % 128 == 0`), so 1152 would still 32-page-split there. We did not measure a CUDA abort; the hard GPU fault we hit is ROCm.

## Suggested fix

Give `RocmPlatform` the same alignment CUDA already uses on sm120:

```python
@classmethod
def _get_indexer_block_alignment(cls, vllm_config: "VllmConfig") -> int | None:
    index_kpool = getattr(
        vllm_config.model_config.hf_text_config, "index_kpool", None
    )
    if not index_kpool or index_kpool <= 1:
        return None
    from vllm.utils.deep_gemm import PAGED_MQA_PAGE_SIZES
    return int(index_kpool) * max(PAGED_MQA_PAGE_SIZES)
```

On this model that turns 1152 → 1280. Then `1280 / 4 = 320` and `320 % 64 == 0` → native **64-entry** pages.

Local boot after that override (same host, MTP still on):

```
Setting attention block size to 1280 tokens
kpool-write: kv_cache shape=(42810, 64, 132) stride=(8448, 132, 1)
             contiguous=True page_size=64
paged-mqa:   KVBlockSize=64 n_blocks=42810 stride0=8448
             i32_limit_block=254200 overflow_layout=False
```

Indexer cache is contiguous (not the DeepSeek-V4 packed stride-1e6 layout). int32 `buffer_load` overflow is not this GLM layout (`overflow_layout=False`). Mamba page padding rose 6.27% → 18.08%; KV pool still ~10.77M tokens at `max_model_len=1000000`.

## Verified

- Hybrid `block_size=1152` → storage 288 → 32-page split, measured on the serving image (2026-09-04).
- Crash B already had the #54296 V2 `in_range` guard and still aborted, so that PR is not this fault.
- After aligning to `index_kpool * max(PAGED_MQA_PAGE_SIZES)`, boot log shows 1280 / `page_size=64` / contiguous. Replay: unique ~134k-token MTP request, plus mixed concurrency=5; **0** `Memory access fault`.
- On current `main` @ `156050598eea`: CUDA helper is `cuda.py:414-431` (sm120 already `max`); `RocmPlatform` still has no override (grep of `rocm.py` at that sha: 0 hits for `_get_indexer_block_alignment` / `PAGED_MQA` / `index_kpool`); `attention.py:145-152` still falls back to 32 when `storage % 64 != 0`.

The local serving image that stopped aborting also carried defensive store-masks / block-id clamps in the kpool write and paged-MQA helpers. Those were **not** A/B-isolated from the alignment change, so this report only claims the layout gap and that the 64-page image no longer faulted on replay.

## Not verified / not claimed

- HIP kernel name of the aborting launch. We did **not** recapture Crash A/B with `HIP_LAUNCH_BLOCKING` on the old tag; MTP decode is the scheduler step, not a named kernel.
- Repro on vanilla `main` (no #53906-era local image). The missing `RocmPlatform` override **is** on `main`; the 1152 hybrid number is from our serving image.
- CUDA abort for the same 1152 → 32-page geometry.
- Whether #54359’s “backend advertises `MultipleOf(16)`, no kernel-granular block table” is required for this abort. Different symptom (that issue is silent overwrite / fluent text; this is a GPU fault).

## Not a duplicate of

- **#54359** (open): ROCm indexer **silently overwrites** its own KV because the backend advertises a kernel block size the kernels do not use. Nothing crashes. This report is a hard `Memory access fault` from the 32-page split of a 1152 hybrid block.
- **#54296** (open): V2 slot-mapping `in_range`. Necessary, and Crash B already had it.
- **#53906** (merged 2026-09-03): GLM-5.3-Flash support, including the CUDA helper above. The ROCm platform still returns `None` for indexer alignment.
- **#52109** and [ROCm/aiter#4774](https://github.com/ROCm/aiter/issues/4774) (closed): DeepSeek-V4 packed indexer / gluon int32 offsets past 2 GiB. Different model and symptom (silent zero logits vs a GPU fault). This GLM indexer is contiguous (`overflow_layout=False`).

Happy to re-run on gfx942 against a PR that adds the `RocmPlatform` override.

**Comment by zzw09773** [comment]
Correction to the root-cause claim in the issue body.

The ROCm indexer-alignment gap is still real on current `main` (`ce6c241ddcc722ee7f26fb4db132461150dad18e`): `RocmPlatform` still does not override `_get_indexer_block_alignment`, so hybrid GLM can stay on a block size that virtually splits to 32-entry pages. That is worth fixing on its own. It is **not** a sufficient explanation of the gfx942 `Memory access fault` family this issue was opened for.

We kept serving on an image that already had the suggested alignment (`block_size=1280`, native `page_size=64`) plus the #54296 V2 `in_range` guard. The abort continued.

**Crash C** (2026-09-04 06:10:46Z), already-aligned image: 3-way all-MTP decode, computed `[129934, 20, 20]`, `new_block_ids=[None,None,None]`, KV 1.56%. Same banner (`GPU node-5` → `VllmWorker-3` exit -6). The long request had already produced 3419 tokens; two short MTP joins landed on the same step. HIP kernel name was **not** captured. Locally we treated this as mixed-length decode into aiter’s non-varctx gluon paged-MQA (inferred, not `HIP_LAUNCH_BLOCKING`).

**Crash D** (2026-09-07 08:18:08Z), same box, image that already had the Crash C decode gate: **not** MTP. Scheduler dump is one chunked prefill, computed **211200**, scheduled **2338**, output 0, `scheduled_spec_decode_tokens={}`, `new_block_ids` around **3535–3578**. Prefix-cache ~94.6%. `2338 % 4 == 2`, so the last chunk must seed the kpool tail ring.

This time we recaptured on the dying tag with `HIP_LAUNCH_BLOCKING=1`: `ShaderName : _kpool_tail_seed_kernel`, then `Memory Fault Error`.

On `main` @ `ce6c241d` that kernel is still ungated (`vllm/models/glm5next/amd/ops/kpool_compress.py:357-398`). It does `blk = t // KPOOL` and stores at `blk * TAIL_BLOCK_ELEMS` with no `blk < tail.shape[0]` check. After prefix-cache churn the slot mapping can carry a main-KV physical id into a 1-block-per-request tail tensor.

So please do not treat this issue as “align to 64-page and close”, and do not treat #54296 as the abort fix (Crash B/C/D already had that guard). The layout helper on ROCm is still missing; the faults we actually died on were later, separate paths (mixed-length decode; prefill tail seed). Happy to turn the tail-seed bound into a small PR if that is useful.
